src/extra/components/timepix.py

```diff
@@ -0,0 +1,612 @@
+
+    @property
+    def raw_tot_key(self):
+        return self.raw_instrument_src['data.tot']
+
+    @property
+    def raw_size_key(self):
+        return self.raw_instrument_src['data.size']
+
+    @property
+    def centroids_control_src(self):
+        if self._centroids_control_src is None:
+            raise ValueError('centroid control data is not available as '
+                             'component was initialized with data not '
+                             'containing it')
+
+        return self._centroids_control_src
+
+    @property
+    def centroids_instrument_src(self):
+        if self._centroids_instrument_src is None:
+            raise ValueError('centroids are not available as component was '
+                             'initialized with data not containing it')
+
+        return self._centroids_instrument_src
+
+    @property
+    def centroids_key(self):
+        return self.centroids_instrument_src['data.centroids']
+
+    @property
+    def centroid_labels_key(self):
+        return self.centroids_instrument_src['data.labels']
+
+    def spatial_bins(self, min_pos=0, max_pos=256, bins_per_px=1):
+        """Build suitable bin edges for Timepix position data.
+
+        The bins are centered around integer pixel position to avoid
+        binning artifacts.
+
+        Args:
+            min_pos (int, optional): Lower spatial boundary,
+                0 by default.
+            max_pos (int, optional): Upper spatial boundary,
+                256 by default.
+            bins_per_px (int, optional): Number of bins per pixel,
+                1 by default
+
+        Returns:
+            bins (numpy.ndarray) Spatial bin edges.
+        """
+
+        return np.arange(min_pos, max_pos, 1/bins_per_px) - 0.5/bins_per_px
+
+    def pixel_events(self, pulse_dim='pulseId', toa_offset=0.0,
+                     timewalk_lut=None, out_of_pulse_events=False,
+                     extended_columns=False, parallel=None):
+        """Get raw pixel events as dataframe.
+
+        Loads the original raw pixel events as a dataframe and sorts
+        them into pulses using the initialized pulse pattern component.
+
+        By default, only events the pulse windows are included. The
+        `out_of_pulse_events` arguments allows to extend this to events
+        before the first and after the last pulse.
+
+        An event is considered part of a particular pulse when its
+        time-of-arrival is between the beginning of that and the next
+        pulse For the last pulse, the single largest spacing between
+        pulses is chosen while in case of a single pulse in the entire
+        train, all pulses will be assigned to it. The `toa_offset`
+        argument denotes the time-of-arrival at the start of the first
+        pulse of the train.
+
+        Args:
+            pulse_dim ('pulseId' or 'pulseIndex' or 'pulseTime', optional):
+                Label for pulse dimension, pulse ID by default.
```

Comment:
Wording nitpick: 'label' sounds like it just changes the... label on the pulse index, whereas this actually changes the information in the index. Also, pandas has 'index levels' instead of dimensions. So what about something like...

```suggestion
                Which pulse metadata to use in the index, pulse ID by default.
```

(I also don't think the difference between pulse ID & index is obvious - I had to think about it - but I don't think we want to explain that in every docstring that refers to them.)

Reply:
I agree this may not always be obvious, and we should explain this in the `PulsePattern` component. But I'm not entirely sure *pulse metadata* is clearer, how about *pulse annotation*?

Reply:
Yup, I'm fine with *annotation*. Or *coordinates*, if you want another option. Thinking about it, even *pulse labels* (plural) might be OK.

Reply:
I like coordinates, thanks. That also fits with train ID coordinates in EXtra-data.